packages/utils/telemetry-utils/src/errorLogging.ts

```diff
@@ -3,102 +3,202 @@
-export function extractLogSafeErrorProperties(error: any) {
-    const removeMessageFromStack = (stack: string, errorName?: string) => {
-        const stackFrames = stack.split("\n");
-        stackFrames.shift(); // Remove "[ErrorName]: [ErrorMessage]"
-        if (errorName !== undefined) {
-            stackFrames.unshift(errorName); // Add "[ErrorName]"
-        }
-        return stackFrames.join("\n");
-    };
-
-    const message = (typeof error?.message === "string")
-        ? error.message as string
-        : String(error);
+/** Metadata to annotate an error object when annotating or normalizing it */
+export interface FluidErrorAnnotations {
+    /** Telemetry props to log with the error */
+    props?: ITelemetryProperties;
+    /** fluidErrorCode to mention if error isn't already an IFluidErrorBase */
+    errorCodeIfNone?: string;
+}
 
-    const safeProps: { message: string; errorType?: string; stack?: string } = {
+/**
+ * Take an unknown error object and extract certain known properties to be included in a new error object.
+ * The stack is preserved, along with any safe-to-log telemetry props.
+ * @param error - An error that was presumably caught, thrown from unknown origins
+ * @param newErrorFn - callback that will create a new error given the original error's message
+ * @returns A new error object "wrapping" the given error
+ */
+export function wrapError<T>(
+    error: any,
+    newErrorFn: (m: string) => T,
+): T {
+    const {
         message,
-    };
+        stack,
+    } = extractLogSafeErrorProperties(error);
+    const props = isILoggingError(error) ? error.getTelemetryProperties() : {};
 
-    if (isRegularObject(error)) {
-        const { errorType, stack, name } = error;
+    const newError = newErrorFn(message);
+    mixinTelemetryProps(newError, props);
 
-        if (typeof errorType === "string") {
-            safeProps.errorType = errorType;
-        }
-
-        if (typeof stack === "string") {
-            const errorName = (typeof name === "string") ? name : undefined;
-            safeProps.stack = removeMessageFromStack(stack, errorName);
-        }
+    if (stack !== undefined) {
+        Object.assign(newError, { stack });
     }
 
-    return safeProps;
+    return newError;
 }
 
-/** Type guard for ILoggingError */
-export const isILoggingError = (x: any): x is ILoggingError => typeof x?.getTelemetryProperties === "function";
+/**
+ * Normalize the given error yielding a valid Fluid Error
+ * @returns The same error object passed in if possible, normalized and with any provided annotations applied
+ * @param error - The error to normalize, ideally by patching in the properties of IFluidErrorBase
+ * @param annotations - Annotations to apply to the normalized error
+ */
+export function normalizeError(
+    error: unknown,
+    annotations: FluidErrorAnnotations = {},
+): IFluidErrorBase {
+    if (isRegularObject(error) && !Object.isFrozen(error)) {
```

Comment:
I'm not sure I like Object.isFrozen check. I feel contract should be more explicit - either caller expects arguments to be mutated by this function, or function should accept readonly objects and just clone them.

Where and why we freeze objects?

Reply:
Hopefully we don't freeze. The purpose of nornalizeError is for anywhere we do `catch`. We need to be able to handle completely arbitrary input, so I'm focused on making this robust.

Reply:
I looked into the cloning readonly objects earlier but wasn't confident about how to implement it so went with this simpler (more strict) approach. We can see in telemetry if this ever happens and deal with it. 

Or if you can share the right code snippet I'd consider doing that approach now.

Reply:
lodash has cloneDeep(), and we even use it in FF (even though I think Curtis does not like it :))
I'm not sure it's worth it, I'd probably simplify and say that we do not support frozen objects at all, i.e. assert objects are not frozen (and change our mind if telemetry tells us otherwise). I feel it's a bit simpler message than saying that one function is Ok with frozen objects and another one is not

Reply:
> say that we do not support frozen objects at all,

Ok yeah, I like that 👌

I mean, who would care to freeze an error? (famous last words)